src/blocker.cr

```diff
@@ -52,10 +53,18 @@ class Castblock::Blocker
 
   private def watch_device(device : Chromecast::Device, continue : Channel(Nil)) : Nil
     @chromecast.start_watcher(device, continue) do |message|
-      Log.debug &.emit("Received message", application: message.application.display_name)
+      if application = message.application
+        Log.debug &.emit("Received message", application: application.display_name)
+        if application.display_name.downcase == "youtube" && (media = message.media) && media.player_state == "PLAYING"
+          handle_media(device, media)
+        end
+      end
 
-      if message.application.display_name.downcase == "youtube" && (media = message.media) && media.player_state == "PLAYING"
-        handle_media(device, media)
+      if @mute_ads && (payload = message.payload)
+        data = JSON.parse(payload)
+        if data["status"]? && (player_state = data["status"][0]["customData"]["playerState"].as_i)
```

Comment:
This bothers me a bit…

The structure is pretty deep, but still I would prefer a `Payload` object (and other nested ones), to avoid using `JSON.parse` and manipulating hashes.

Reply:
Yep I know it's ugly. Part of the difficulty is that there are 2 types of updates from go-chromecast watch, the one that comes at a 1 or 2 second interval, and automatic messages from the chromecast. Some (or all?) of the automatic updates come through the payload param, which I believe hasn't been used until now seeing that it isn't in the WatchMessage struct.

The other difficulty is that go-chromecast supplies the payload param as a json string, and I couldn't figure out how I would put in a struct that the json needs to be parsed into the struct, then the payload param needs to be parsed again into another struct since it was a json string embedded in json. I gave it a shot but I've never written any crystal before so I did it the hacky way.

This might be an ask for Vishen to change upstream to actually put the payload into the json returned by go-chromecast - or maybe the payload could be parsed and relayed in a similar format as the regular (1 or 2 second) updates.

At any rate, using the payload data seems to be the best way to get a low enough latency to not cut off the beginning of videos.

Reply:
> The other difficulty is that go-chromecast supplies the payload param as a json string

This is indeed not great :confused: